src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -144,24 +144,28 @@ SHAMap::dirtyUp (SharedPtrNodeStack& stack,
     }
 }
 
-SHAMapTreeNode* SHAMap::walkToPointer (uint256 const& id) const
+SHAMapTreeNode*
+SHAMap::walkToKey(uint256 const& id, NodeStack* stack) const
 {
+    assert(stack == nullptr || stack->empty());
     auto inNode = root_.get();
     SHAMapNodeID nodeID;
+    if (stack != nullptr)
+        stack->push({inNode, nodeID});
 
-    while (inNode->isInner ())
+    while (inNode->isInner())
     {
-        int branch = nodeID.selectBranch (id);
-        auto inner = static_cast<SHAMapInnerNode*>(inNode);
+        auto const branch = nodeID.selectBranch (id);
+        auto const inner = static_cast<SHAMapInnerNode*>(inNode);
         if (inner->isEmptyBranch (branch))
             return nullptr;
 
         inNode = descendThrow (inner, branch);
         nodeID = nodeID.getChildNodeID (branch);
+        if (stack != nullptr)
+            stack->push({inNode, nodeID});
     }
-
-    auto ret = static_cast<SHAMapTreeNode*>(inNode);
-    return ret->peekItem()->key() == id ? ret : nullptr;
+    return static_cast<SHAMapTreeNode*>(inNode);
```

Comment:
@HowardHinnant I agree we want to keep a function like `upper_bound` efficient. Since it's checked in four places, I think that enough to justify a wrapper.